### ISSUE TITLE
Only emit results on success

### ISSUE THREAD

**Issue by cirocosta** [vmware-tanzu/cartographer#4]
As a pipeline author
I want the status of my Pipeline to be based on the _most recently submitted_ and _successful_ PipelineRun 
So that I can trigger work only when a test passes

# Basic Spec
```gherkin
Given a pipeline and runTemplate that have created an object
When that object has a Succeeded condition that is true
Then the Pipeline's Status includes the output fields defined on the runTemplate
```

```gherkin
Given a pipeline and runTemplate that has created an object
When that object has a Succeeded condition that is not true
Then the Pipeline's Status is still that of a previous successful run
```

# Notes:
## Hard coding 'success'
`Success` has been scoped to:
When `stampedObject.Status.Conditions[@type='succeeded'].Status` == true 

## Compare based on creationTimestamp
1. We expect one pipeline can kick off multiple stampedobjects at once.
2. Those objects can succeed in indeterminate order.
3. We expect to emit status from the most recently submitted object which is successful.
4. In this story, most recently successful will be judged by their creationTimestamp.

# Out of scope:
## Ordering objects defined more than once
This scheme has a corner case for when a pipeline has spec A (creates an object), is updated to spec B (creates another object), and then is updated to spec A (will not always create a new object because the original exists). In that case, when both A and B are successful, the most recently submitted successful object should be A (it is so recent that it is the current spec of the pipeline). The current story need not handle this corner case.

## Time collisions
One could imagine a scenario where 1 pipeline submitted two objects in such close succession that they have the same creationTimestamp. In that scenario, it would be impossible to judge which is the most recently submitted. This story need not handle that case. There may be a follow-on story to create some mechanism to move from timestamps to some monotonically increasing series. This series would need to be appended to the objects created by the pipeline. Story authors should remember that new object creation can be triggered by changes to the pipeline, the runTemplate or (possibly) some 3rd object to which they refer.

**Comment by DanieldeR** [comment]
As discussed at IPM: 

* We should clarify the terminology around the overloaded "pipeline" term as it's used for both the pipeline service as a Cartographer construct and throughout Tekton. (@DanieldeR to pick this up) 
* There is a story missing that would implement a blocking mechanism that would stop Cartographer from creating new objects if currently "running" objects were in flight. This second story should have a milestone of beta. (@waciumawanjohi can you creat this story?)
  * The race condition of this issue is no longer necessary if this second story is created.

**Comment by squeedee** [comment]
Have some concerns, even with sequential builds:

Pipeline
```yaml
metadata:
  generation: 1
spec:
  inputs:
    revision: 11
```

triggers a stampedObject and we get

```yaml
metadata:
  generation: 1
spec:
  inputs:
    revision: 11
status:
  observedGeneration: 1
  conditions:
    - type: Ready
      status: Unknown
    - type: StampedObjectReady
      status: Unknown
```

while first stamp is running, a new revision drops

```yaml
metadata:
  generation: 2
spec:
  inputs:
    revision: 12
status:
  observedGeneration: 1 # current condition logic would increment this. we need to ensure the status matches the object stamped for the first generation
  conditions:
    - type: Ready
      status: Unkown
    - type: StampedObjectReady
      status: Unknown
```

And the stampedObject completes successfully:

```yaml
metadata:
  generation: 2
spec:
  inputs:
    revision: 12
status:
  observedGeneration: 1 
  conditions:
    - type: Ready
      status: True
    - type: StampedObjectReady
      status: True
```

 ** But we wont have this object for very long?! **

The next reconcile gives us:


```yaml
metadata:
  generation: 2
spec:
  inputs:
    revision: 12
status:
  observedGeneration: 2 
  conditions:
    - type: Ready
      status: Unknown
    - type: StampedObjectReady
      status: Unknown
```

**Comment by jwntrs** [comment]
@squeedee does it matter that we don't have that object for long? The SupplyChain will likely only care about the `latest(Successful)Outputs` right?

**Comment by squeedee** [comment]
@jwntrs so if person-a merges a good PR, and right after person-b merges a bad (will fail the pipeline) PR - it wont bother person-a that their good, and positively tested PR did not end up producing an image because the person-b pr failed?

**Comment by squeedee** [comment]
Or is the idea that we don't use "Unknown" but stick to 

```yaml
status:
  observedGeneration: 1 
  conditions:
    - type: Ready
      status: True
  outputs:
      ....
```

until such time as generation 2 is complete? 

Upside, we only present good outputs matched with the spec generation that created them. Downside, we haven't indicated that there is now a new stamped object busily performing a pipelineRun (for example)

**Comment by squeedee** [comment]
We should implement #116 first.

@jwntrs Talking with @waciumawanjohi it seems that the best solution to my above concerns is to have:

```yaml
status:
  latestSuccessful:
    outputs:
       ...
```

These would only be populated from the most recent stampedObject with conditions:
```yaml
status:
  conditions:
    - type: Ready
      status: True
```

**Comment by jwntrs** [comment]
@squeedee since there is no guarantee that the GitRepository will even pick up a particular commit, I'm not sure that we can guarantee much of anything other than the fact that we will eventually get the `latest thing`.

Do we even need #116? Can't we just take the latest `Ready - True` based on creationTimestamp?

**Comment by waciumawanjohi** [comment]
Removed the `race condition` section from the spec and updated the story to reflect engineering conversations and agreements. Duplicating the original race condition language here for posterity:

# Race condition Spec

If we run concurrently, we must pick the very latest completion (not the last submitted run).
Conversely, we could choose to not run until there are no running children (and so don't start a new run until the last one completed)

```gherkin
Given a Tekton pipeline run template
When I apply a Pipeline that will pass, but very slowly
Then I see the PipelineRun has started a tekton pipeline
When I apply a pipeline that will not pass, but runs quickly
And the quick PipelineRun completes
Then I see the Pipeline's Status is invalid
When the slow PipelineRun completes
Then I see the Pipeline's Status is that of the successful run
```
